src/llm_models/perplexity.rs

```diff
@@ -0,0 +1,161 @@
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use log::info;
+use reqwest::{header, Client};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+
+use crate::constants::PERPLEXITY_API_URL;
+use crate::domain::{PerplexityAPICompletionsResponse, RateLimit};
+use crate::llm_models::LLMModel;
+use crate::utils::{map_to_range_f32, sanitize_json_response};
+
+#[derive(Deserialize, Serialize, Debug, Clone, Eq, PartialEq)]
+//Mistral docs: https://docs.mistral.ai/platform/endpoints
+pub enum PerplexityModels {
+    Llama3_1SonarSmall,
+    Llama3_1SonarLarge,
+    Llama3_1SonarHuge,
+}
+
+#[async_trait(?Send)]
+impl LLMModel for PerplexityModels {
+    fn as_str(&self) -> &str {
+        match self {
+            PerplexityModels::Llama3_1SonarSmall => "llama-3.1-sonar-small-128k-online",
+            PerplexityModels::Llama3_1SonarLarge => "llama-3.1-sonar-large-128k-online",
+            PerplexityModels::Llama3_1SonarHuge => "llama-3.1-sonar-huge-128k-online",
+        }
+    }
+
+    fn try_from_str(name: &str) -> Option<Self> {
+        match name.to_lowercase().as_str() {
+            "llama-3.1-sonar-small-128k-online" => Some(PerplexityModels::Llama3_1SonarSmall),
+            "llama-3.1-sonar-large-128k-online" => Some(PerplexityModels::Llama3_1SonarLarge),
+            "llama-3.1-sonar-huge-128k-online" => Some(PerplexityModels::Llama3_1SonarHuge),
+            _ => None,
+        }
+    }
+
+    // TODO: https://docs.perplexity.ai/guides/rate-limits
+    fn default_max_tokens(&self) -> usize {
+        127_072
+    }
+
+    fn get_endpoint(&self) -> String {
+        PERPLEXITY_API_URL.to_string()
+    }
+
+    //This method prepares the body of the API call for different models
+    fn get_body(
+        &self,
+        instructions: &str,
+        json_schema: &Value,
+        function_call: bool,
+        // The total number of tokens requested in max_tokens plus the number of prompt tokens sent in messages must not exceed the context window token limit of model requested.
+        // If left unspecified, then the model will generate tokens until either it reaches its stop token or the end of its context window.
+        _max_tokens: &usize,
```

Comment:
Because Perplexity adds the prompt tokens and requested response when calculating the token limit, it is better to not to pass anything as max_tokens and use default instead. Otherwise we would have to calculate prompt tokens etc.